File: C++/Polymorphism.cpp
#include <iostream>
#include <string>
using namespace std;

class NoiseMaker {
public:
	virtual string makeNoise() =0;
};

class Sheep : public NoiseMaker {
public:
	string makeNoise() {return "Beeeeiii\n";}
};

class Car : public NoiseMaker {
public:
	string makeNoise() {return "Vrooom\n";}
};

void print(NoiseMaker* m) {
	cout << m->makeNoise();
}

int main(int argc, char const *argv[]) {
	Car* c = new Car();
	Sheep* s = new Sheep();

	print(c);
	print(s);

	return 0;
}

File: objC/Polymorphism.m
#import <Foundation/Foundation.h>

@protocol noiseProtocol
- (NSString*) makeNoise;
@end

@interface Sheep : NSObject <noiseProtocol>
@end

@implementation Sheep
- (NSString*) makeNoise {return @"Beeeeiii\n";}
@end

@interface Car : NSObject <noiseProtocol>
@end

@implementation Car
- (NSString*) makeNoise {return @"Vrooom\n";}
@end

print(id<noiseProtocol> noiseMaker) {
	printf([[noiseMaker makeNoise] UTF8String]);
}

int main(int argc, char const *argv[]) {
	print([Sheep alloc]);
	print([Car alloc]);
	return 0;
}